orbmain/src/main/java/com/sun/corba/ee/impl/javax/rmi/CORBA/Util.java

```diff
@@ -195,16 +195,48 @@ public RemoteException mapSystemException(SystemException ex)
             RemoteException newEx = new NoSuchObjectException(message);
             newEx.detail = ex;
             return newEx;
-        } else if (ex instanceof TRANSACTION_REQUIRED) {
-            RemoteException newEx = new TransactionRequiredException(message);
-            newEx.detail = ex;
-            return newEx;
-        } else if (ex instanceof TRANSACTION_ROLLEDBACK) {
-            RemoteException newEx = new TransactionRolledbackException(message);
-            newEx.detail = ex;
-            return newEx;
-        } else if (ex instanceof INVALID_TRANSACTION) {
-            RemoteException newEx = new InvalidTransactionException(message);
+        } else if (ex instanceof TRANSACTION_REQUIRED || ex instanceof TRANSACTION_ROLLEDBACK || ex instanceof INVALID_TRANSACTION) {
+            RemoteException newEx;           
+            Class<RemoteException> clazz = null;
+            if (ex instanceof TRANSACTION_REQUIRED) {
+                try {
+                    clazz = (Class<RemoteException>) Class.forName("javax.transaction.TransactionRequiredException", true, Thread.currentThread().getContextClassLoader());
```

Comment:
The basic logic of selecting which exception is thrown is the same for each pair, but is harder to read than necessary. Please extract it into a method that takes a list of possible exception names to create, returning the first that successfully instantiates.